### DISCUSSION TITLE
Segment.com tracking snippet, URL parameters

### DISCUSSION THREAD

**Discussion by rashad** [vercel/next.js#31958] [Help] [answered]
Hello everyone, 

First of all thanks a lot for all the great work here ! 
I've been using the segment snippet in a nextjs application and everything is going well. (I followed [nextjs official integration process](https://github.com/vercel/next.js/tree/canary/examples/with-segment-analytics))
However, I've found myself trying to send some url parameters as event properties in the default `page` call for 2 main reasons
1. having many urls like `myproject.com/foo&label=ABC` and `myproject.com/foo&label=DEF` makes it impossible to analyze the data without struggle or ETL
2. it breaks Amplitude's quota of event types number since i get a new one every time a new event combinaison is created 😅 

Any idea how I can get out of this ?

Cheers 👋

**Comment by jamesmosier** [comment, marked as answer]
> trying to send some url parameters as event properties

I assume you are talking about wanting to send new event properties here:

```js
Router.events.on('routeChangeComplete', (url) => {
  window.analytics.page(url, { property1: 123 })
})
```

You could parse the `url` value in order to get the query string values you want (there are many suggestions on that [here](https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript)). Additionally you could remove them from `url` if you are wanting to ensure that the url you send to segment doesn't include any extra query string values.

The [`Router.event.on` function](https://nextjs.org/docs/api-reference/next/router#routerevents) only contains the full `url` from the browser, so unfortunately it doesn't include any built-in properties other than that.

**Reply by rashad** [reply]
Simply perfect ! Thanks a lot for your help ! The suggestions you gave me put me on the right direction.

Here's what I did for future folks who might encounter the same problem.
Use the event suggested by @jamesmosier 
```js
Router.events.on('routeChangeComplete', () => {
    const urlSearchParams = new URLSearchParams(window.location.search);
    const params = Object.fromEntries(urlSearchParams.entries());
    window.analytics.page(params)
})
```
then here's an example of an event I get in Segment
```js
// real event with fake id
analytics.page({
  labelId: 'vQD3oUTzYRJAzG4',
  leagueId: 'vQD3oUTzYRJAzG4',
  path: '/rapper',
  rapperId: 'vQD3oUTzYRJAzG4',
  referrer: 'http://localhost:3000/market?leagueId=vQD3oUTzYRJAzG4&labelId=vQD3oUTzYRJAzG4&searchMode=true',
  search: '?labelId=vQD3oUTzYRJAzG4&leagueId=vQD3oUTzYRJAzG4&rapperId=vQD3oUTzYRJAzG4',
  title: '',
  url: 'http://localhost:3000/rapper?labelId=vQD3oUTzYRJAzG4&leagueId=vQD3oUTzYRJAzG4&rapperId=vQD3oUTzYRJAzG4'
});
```

⚠️ to avoid event duplication in Segment, do not forget to turn of the automatic page call from the `_document.js` snippet like
```js
renderSnippet() {
    const opts = {
      apiKey: ANALYTICS_WRITE_KEY,
      page: false,  //<<- HERE
    }

    if (NODE_ENV === 'development') {
      return snippet.max(opts)
    }

    return snippet.min(opts)
  }
```
that is used in the `Head` component like
```js
{/* Inject the Segment snippet into the <head> of the document  */}
<script dangerouslySetInnerHTML={{ __html: this.renderSnippet() }} />
```

Which is exactly what I wanted ! 
Thanks again folks 👋 🇫🇷
